File: 2015/day02.py
from typing import Callable
from os.path import isfile, join as path_join
file_name = path_join('input', 'day02.txt')
def to_list(mf: Callable = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: Callable = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file()

presents = [list(map(int, line.split("x"))) for line in open(file_name).read().strip().split("\n")]

def calc_surface_area(length, width, height):
    return (2 * length * width) + (2 * width * height) + (2 * height * length)


def calc_slack_area(length, width, height):
    sides_1 = length * width
    sides_2 = width * height
    sides_3 = height * length

    return min(sides_1, sides_2, sides_3)

def part1():
    total_area = 0

    for present in presents:
        total_area += calc_surface_area(present[0], present[1], present[2])
        total_area += calc_slack_area(present[0], present[1], present[2])

    return total_area


def calc_ribbon_length(length, width, height):
    perim_1 = 2 * length + 2 * width
    perim_2 = 2 * width + 2 * height
    perim_3 = 2 * length + 2 * height

    return min(perim_1, perim_2, perim_3)


def calc_bow_length(length, width, height):
    return length * width * height


def part2():
    ribbon_length = 0

    for present in presents:
        ribbon_length += calc_ribbon_length(present[0], present[1], present[2])
        ribbon_length += calc_bow_length(present[0], present[1], present[2])

    return ribbon_length


print(part1())
print(part2())

File: cpp/2015/day02.cpp
#include "../aocHelper.h"

class Day02 : public BaseDay
{
public:
	Day02() : BaseDay("02") {}

	result_type solve() override
	{
		int part1 = 0, part2 = 0;

		auto surface_area = [](int l, int w, int h)
		{
			return 2 * ((l * w) + (w * h) + (h * l));
		};

		auto slack_area = [](int l, int w, int h)
		{
			return min(l, min(w, h));
		};

		auto ribbon_length = [](int l, int w, int h)
		{
			return 2 * min(l + w, min(w + h, l + h));
		};
		
		auto bow_length = [](int l, int w, int h)
		{
			return l * w * h;
		};

		while (*input != '\0')
		{
			int length = numericParse<int>(input);
			input += 2; // skip: x
			int width = numericParse<int>(input);
			input += 2; // skip: x
			int height = numericParse<int>(input);

			// part 1
			part1 += surface_area(length, width, height);
			part1 += slack_area(length, width, height);

			// part 2
			part2 += ribbon_length(length, width, height);
			part2 += bow_length(length, width, height);

			input += 2; //skip: \n
		}

		return { part1, part2 };
	}
};
